### PR TITLE
add weighted random mint and referrer

### PR DESCRIPTION
**Pull request by jillxuu**

### CODE REVIEW

File: airdrop-machine/sources/airdrop_machine.move [hunk 1]
```diff
@@ -142,7 +164,7 @@ module airdrop_machine::airdrop_machine {
         let collection_owner_signer = collection_owner_signer(collection_config);
         let metadata_config = collection_config.metadata_config;
         let collection = collection_config.collection;
-        let index = get_pseudo_random_index(vector::length(&metadata_config.token_uris));
+        let index = get_weighted_pseudo_random_index(&metadata_config.token_uris_weights);
         let uri = *vector::borrow(&metadata_config.token_uris, (index as u64));
```
Comment (JohnChangUK) [inline airdrop-machine/sources/airdrop_machine.move:168]: `index` doesn't need to be casted right? `index` is already `u64`

File: airdrop-machine/sources/airdrop_machine.move [hunk 2]
```diff
@@ -278,9 +302,41 @@ module airdrop_machine::airdrop_machine {
...
+    fun get_weighted_pseudo_random_index(token_uris_weights: &vector<u64>): u64 {
+        let total_weight = 0u64;
+        let len = vector::length(token_uris_weights);
+
+        // Compute the total weight
+        let i = 0;
+        while (i < len) {
+            total_weight = total_weight + *vector::borrow(token_uris_weights, i);
+            i = i + 1;
+        };
```
Comment (JohnChangUK) [inline airdrop-machine/sources/airdrop_machine.move:314]: We can replace all this with

`let total_weight = vector::fold(*token_uris_weights, 0, |l, r| l + r);`

File: airdrop-machine/sources/airdrop_machine.move [hunk 3]
```diff
@@ -278,9 +302,41 @@ module airdrop_machine::airdrop_machine {
...
         let txn_hash = transaction_context::get_transaction_hash();
-        ((*vector::borrow(&txn_hash, 0) as u64) * 256u64) % length
+        let seed = 0u64;
+        let j = 0;
+        while (j < 4 && j < vector::length(&txn_hash)) {
+            let byte = (*vector::borrow(&txn_hash, j) as u64);
+            seed = seed + (byte << ((j * 8) as u8));
+            j = j + 1;
+        };
+        let random_weight = seed % total_weight;
```
Comment (JohnChangUK) [inline airdrop-machine/sources/airdrop_machine.move:325]: Really neat way of generating a seed!

File: airdrop-machine/sources/airdrop_machine.move [hunk 4]
```diff
@@ -29,22 +30,33 @@ module airdrop_machine::airdrop_machine {
     /// Token Minting has not yet started.
     const EMINTING_HAS_NOT_STARTED: u64 = 3;
     
-    struct MetadataConfig has store, copy, drop {
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
```
Comment (JohnChangUK) [inline airdrop-machine/sources/airdrop_machine.move:33]: If this is an object, perhaps we should store this resource in the same object signer as CollectionConfig. And when fetching it, use `borrow_global<MetadataConfig>(object::object_address(&collection_config_object);`

We can then remove `store, copy` drop` capabilities too from the struct.

File: airdrop-machine/sources/airdrop_machine.move [hunk 5]
```diff
@@ -278,9 +302,41 @@ module airdrop_machine::airdrop_machine {
...
+            total_weight = total_weight + *vector::borrow(token_uris_weights, i);
+            i = i + 1;
+        };
+
+        // Obtain a pseudo-random value using the transaction hash
         let txn_hash = transaction_context::get_transaction_hash();
-        ((*vector::borrow(&txn_hash, 0) as u64) * 256u64) % length
+        let seed = 0u64;
+        let j = 0;
+        while (j < 4 && j < vector::length(&txn_hash)) {
```
[the remark this answers is not in the document]
Comment (JohnChangUK) [reply airdrop-machine/sources/airdrop_machine.move:320]: Should probably have a comment on what the code is doing too. I.e, parsing through the TX hash (32 bytes), going through 8 bytes at a time.

File: airdrop-machine/sources/airdrop_machine.move [hunk 6]
```diff
@@ -29,22 +30,33 @@ module airdrop_machine::airdrop_machine {
...
     struct CollectionConfig has key {
         metadata_config: MetadataConfig,
         collection: Object<Collection>,
         extend_ref: object::ExtendRef,
         ready_to_mint: bool,
     }
 
+    #[event]
+    /// Emitted when a token has been minted by a user.
+    struct Mint has drop, store {
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply airdrop-machine/sources/airdrop_machine.move:54]: yes, but emitting events would be much easier for future aggregation

File: airdrop-machine/sources/airdrop_machine.move (hunk 4 above)
Comment (jillxuu) [reply airdrop-machine/sources/airdrop_machine.move:33]: after thinking more, i don't see the necessity of having to use objectgroup actually.

**Review by JohnChangUK — APPROVED** [review]

**Pull request merged**